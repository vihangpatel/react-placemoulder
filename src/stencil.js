import React, { useRef, useEffect, useMemo } from 'react';

import {
  mouldParagraph,
  createObjectFromSchema,
  createDummyArray,
} from './utils';

import './stencil.css';

export const Stencil = ({ children, ...rest }) => {
  const currentRef = useRef(null);

  useEffect(() => {
    mouldParagraph(
      Array.prototype.slice.call(
        currentRef.current.querySelectorAll('.stensil-para')
      )
    );
  }, []);

  return (
    <div
      ref={currentRef}
      tabIndex="-1"
      aria-disabled="true"
      {...rest}
      className={`enable-stensil ${rest.className || ''}`.trim()}
    >
      {children}
    </div>
  );
};

export const StencilList = ({
  data,
  length,
  schema,
  Component,
  children,
  ...rest
}) => {
  // Create dummy props if config is provided
  const dummyProps = data || createObjectFromSchema(schema);

  // Memoize fakeArray based on length
  const fakeArray = useMemo(() => createDummyArray(length), [length]);

  return (
    <Stencil {...rest}>
      {Component
        ? fakeArray.map((index) => (
            <Component key={index} {...dummyProps} stencilLoading />
          ))
        : null}
    </Stencil>
  );
};

export const StencilWrapper = ({ repeat, children, ...rest }) => {
  // Memoize fakeArray based on length
  const fakeArray = useMemo(() => createDummyArray(repeat), [repeat]);

  const isOnlyChild = React.Children.only(children);

  return (
    <Stencil {...rest}>
      {fakeArray.map((key) =>
        React.cloneElement(children, { stencilLoading: true, key })
      )}
    </Stencil>
  );
};
