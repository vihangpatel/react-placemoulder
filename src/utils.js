const REAPEAT_CHAR = ' .';

const noop = () => {};

export const createObjectFromSchema = (schema) => {
  const outputObject = {};

  for (const key in schema) {
    const val = schema[key];

    /** Ignore undefined or null keys and continue */
    if (typeof val === 'undefined' || val === null) {
      continue;
    }

    switch (typeof val) {
      case 'string':
      case 'number':
        outputObject[key] = REAPEAT_CHAR.repeat(+val);
        break;

      case 'function':
        outputObject[key] = noop;
        break;

      case 'boolean':
        outputObject[key] = val;
        break;

      case 'object':
        outputObject[key] = createObjectFromSchema(val);

      default:
        break;
    }
  }

  return outputObject;
};

export const mouldParagraph = (pTags) =>
  pTags.map(
    (pTag) =>
      (pTag.innerHTML = `${Array.apply(null, Array(4)).reduce(
        (resultStr, item) =>
          `${resultStr}<span class="stensil">${' .'.repeat(50)}</span>`,
        ''
      )}<span class="stensil half"/>`)
  );

export function createDummyArray(length) {
  return Array.apply(null, Array(length)).map(function (item, index) {
    return index;
  });
}
