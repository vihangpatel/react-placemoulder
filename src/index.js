import { Stencil, StencilList, StencilWrapper } from './stencil';
import { createObjectFromSchema } from './utils';

export { Stencil, StencilList, StencilWrapper, createObjectFromSchema };
