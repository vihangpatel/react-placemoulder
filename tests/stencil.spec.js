import React from 'react';

import DummyComponent, { props } from './dummy-component';

import { Stencil, StencilList, StencilWrapper } from '../src';

describe('Test react-placemoulder', () => {
  describe('Test <Stencil />', () => {
    it('should render base structure', () => {
      const wrapper = mount(
        <Stencil>
          <DummyComponent />
        </Stencil>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Test <StencilList />', () => {
    it('should test base structure', () => {
      const wrapper = mount(
        <StencilList
          length={3}
          Component={DummyComponent}
          schema={{ name: 12, lastSeen: 12, lastChat: 72 }}
        >
          <DummyComponent data={props} />
          <DummyComponent data={props} />
        </StencilList>
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render empty shell if Component is not passed', () => {
      const wrapper = mount(
        <StencilList
          length={3}
          schema={{ name: 12, lastSeen: 12, lastChat: 72 }}
        ></StencilList>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Test <StencilWrapper />', () => {
    it('should render base structure', () => {
      const wrapper = mount(
        <StencilWrapper repeat={3} className="vertical-list">
          <DummyComponent data={props} />
        </StencilWrapper>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
