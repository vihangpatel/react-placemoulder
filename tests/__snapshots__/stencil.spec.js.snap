// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test react-placemoulder Test <Stencil /> should render base structure 1`] = `
<Stencil>
  <div
    aria-disabled="true"
    className="enable-stensil"
    tabIndex="-1"
  >
    <WhatsAppCard>
      <div
        className="whatsapp-card"
      >
        <div
          className="whatsapp-avatar stensil"
        >
          <img
            className="stensil-ignore"
          />
        </div>
        <div
          className="whatsapp-chat-frame"
        >
          <span
            className="whatsapp-name stensil"
          />
          <p
            className="whatsapp-last-chat stensil-para"
          />
        </div>
        <div
          className="whatsapp-right-block"
        >
          <div
            className="stensil"
          />
        </div>
      </div>
    </WhatsAppCard>
  </div>
</Stencil>
`;

exports[`Test react-placemoulder Test <StencilList /> should render empty shell if Component is not passed 1`] = `
<StencilList
  length={3}
  schema={
    Object {
      "lastChat": 72,
      "lastSeen": 12,
      "name": 12,
    }
  }
>
  <Stencil>
    <div
      aria-disabled="true"
      className="enable-stensil"
      tabIndex="-1"
    />
  </Stencil>
</StencilList>
`;

exports[`Test react-placemoulder Test <StencilList /> should test base structure 1`] = `
<StencilList
  Component={[Function]}
  length={3}
  schema={
    Object {
      "lastChat": 72,
      "lastSeen": 12,
      "name": 12,
    }
  }
>
  <Stencil>
    <div
      aria-disabled="true"
      className="enable-stensil"
      tabIndex="-1"
    >
      <WhatsAppCard
        key="0"
        lastChat=" . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ."
        lastSeen=" . . . . . . . . . . . ."
        name=" . . . . . . . . . . . ."
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              alt=" . . . . . . . . . . . ."
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            >
               . . . . . . . . . . . .
            </span>
            <p
              className="whatsapp-last-chat stensil-para"
            >
               . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
            </p>
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            >
               . . . . . . . . . . . .
            </div>
          </div>
        </div>
      </WhatsAppCard>
      <WhatsAppCard
        key="1"
        lastChat=" . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ."
        lastSeen=" . . . . . . . . . . . ."
        name=" . . . . . . . . . . . ."
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              alt=" . . . . . . . . . . . ."
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            >
               . . . . . . . . . . . .
            </span>
            <p
              className="whatsapp-last-chat stensil-para"
            >
               . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
            </p>
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            >
               . . . . . . . . . . . .
            </div>
          </div>
        </div>
      </WhatsAppCard>
      <WhatsAppCard
        key="2"
        lastChat=" . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ."
        lastSeen=" . . . . . . . . . . . ."
        name=" . . . . . . . . . . . ."
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              alt=" . . . . . . . . . . . ."
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            >
               . . . . . . . . . . . .
            </span>
            <p
              className="whatsapp-last-chat stensil-para"
            >
               . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
            </p>
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            >
               . . . . . . . . . . . .
            </div>
          </div>
        </div>
      </WhatsAppCard>
    </div>
  </Stencil>
</StencilList>
`;

exports[`Test react-placemoulder Test <StencilWrapper /> should render base structure 1`] = `
<StencilWrapper
  className="vertical-list"
  repeat={3}
>
  <Stencil
    className="vertical-list"
  >
    <div
      aria-disabled="true"
      className="enable-stensil vertical-list"
      tabIndex="-1"
    >
      <WhatsAppCard
        data={
          Object {
            "lastChat": "All Good",
            "lastSeen": "12:30 PM",
            "name": "Narendra",
          }
        }
        key="0"
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            />
            <p
              className="whatsapp-last-chat stensil-para"
            />
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            />
          </div>
        </div>
      </WhatsAppCard>
      <WhatsAppCard
        data={
          Object {
            "lastChat": "All Good",
            "lastSeen": "12:30 PM",
            "name": "Narendra",
          }
        }
        key="1"
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            />
            <p
              className="whatsapp-last-chat stensil-para"
            />
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            />
          </div>
        </div>
      </WhatsAppCard>
      <WhatsAppCard
        data={
          Object {
            "lastChat": "All Good",
            "lastSeen": "12:30 PM",
            "name": "Narendra",
          }
        }
        key="2"
        stencilLoading={true}
      >
        <div
          className="whatsapp-card"
        >
          <div
            className="whatsapp-avatar stensil"
          >
            <img
              className="stensil-ignore"
            />
          </div>
          <div
            className="whatsapp-chat-frame"
          >
            <span
              className="whatsapp-name stensil"
            />
            <p
              className="whatsapp-last-chat stensil-para"
            />
          </div>
          <div
            className="whatsapp-right-block"
          >
            <div
              className="stensil"
            />
          </div>
        </div>
      </WhatsAppCard>
    </div>
  </Stencil>
</StencilWrapper>
`;
